File: billing-application/src/main/java/com/example/billing/application/service/ReplayWebhookDeliveryService.java
package com.example.billing.application.service;

import com.example.billing.application.exception.WebhookDeliveryNotFoundException;
import com.example.billing.application.port.in.ReplayWebhookDeliveryUseCase;
import com.example.billing.application.port.out.WebhookDeliveryRepository;
import com.example.billing.domain.webhook.WebhookDelivery;
import com.example.billing.domain.webhook.WebhookDeliveryId;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Clock;

@Service
@RequiredArgsConstructor
@Slf4j
public class ReplayWebhookDeliveryService implements ReplayWebhookDeliveryUseCase {

    private final WebhookDeliveryRepository deliveries;
    private final Clock clock;

    @Override
    @Transactional
    public void replay(WebhookDeliveryId deliveryId) {
        WebhookDelivery delivery = deliveries.findById(deliveryId)
                .orElseThrow(() -> new WebhookDeliveryNotFoundException(deliveryId));
        delivery.replay(clock);
        deliveries.save(delivery);
        log.info("webhook delivery replayed id={} attemptsSoFar={}", deliveryId, delivery.attemptCount());
    }
}

File: billing-application/src/main/kotlin/com/example/billing/application/service/ReplayWebhookDeliveryService.kt
package com.example.billing.application.service

import com.example.billing.application.exception.WebhookDeliveryNotFoundException
import com.example.billing.application.port.`in`.ReplayWebhookDeliveryUseCase
import com.example.billing.application.port.out.WebhookDeliveryRepository
import com.example.billing.domain.webhook.WebhookDeliveryId
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Clock

@Service
open class ReplayWebhookDeliveryService(
    private val deliveries: WebhookDeliveryRepository,
    private val clock: Clock,
) : ReplayWebhookDeliveryUseCase {

    @Transactional
    override fun replay(deliveryId: WebhookDeliveryId) {
        val delivery = deliveries.findById(deliveryId)
            .orElseThrow { WebhookDeliveryNotFoundException(deliveryId) }
        delivery.replay(clock)
        deliveries.save(delivery)
        log.info("webhook delivery replayed id={} attemptsSoFar={}", deliveryId, delivery.attemptCount)
    }

    companion object {
        private val log = LoggerFactory.getLogger(ReplayWebhookDeliveryService::class.java)
    }
}
